File: solutions/javascript/linked-list/1/linked-list.js
export class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this._count = 0;
  }

  // Add element to the end
  push(value) {
    const newNode = new Node(value);
    
    if (this.tail === null) {
      // Empty list
      this.head = newNode;
      this.tail = newNode;
    } else {
      // Add to the end
      this.tail.next = newNode;
      newNode.prev = this.tail;
      this.tail = newNode;
    }
    
    this._count++;
  }

  // Remove and return element from the end
  pop() {
    if (this.tail === null) {
      return undefined; // or throw error
    }
    
    const value = this.tail.value;
    
    if (this.head === this.tail) {
      // Only one element
      this.head = null;
      this.tail = null;
    } else {
      // Remove last element
      this.tail = this.tail.prev;
      this.tail.next = null;
    }
    
    this._count--;
    return value;
  }

  // Remove and return element from the beginning
  shift() {
    if (this.head === null) {
      return undefined; // or throw error
    }
    
    const value = this.head.value;
    
    if (this.head === this.tail) {
      // Only one element
      this.head = null;
      this.tail = null;
    } else {
      // Remove first element
      this.head = this.head.next;
      this.head.prev = null;
    }
    
    this._count--;
    return value;
  }

  // Add element to the beginning
  unshift(value) {
    const newNode = new Node(value);
    
    if (this.head === null) {
      // Empty list
      this.head = newNode;
      this.tail = newNode;
    } else {
      // Add to the beginning
      newNode.next = this.head;
      this.head.prev = newNode;
      this.head = newNode;
    }
    
    this._count++;
  }

  // Delete first occurrence of a value
  delete(value) {
    let current = this.head;
    
    while (current !== null) {
      if (current.value === value) {
        // Found the node to delete
        
        if (current === this.head && current === this.tail) {
          // Only node in the list
          this.head = null;
          this.tail = null;
        } else if (current === this.head) {
          // Deleting head
          this.head = current.next;
          this.head.prev = null;
        } else if (current === this.tail) {
          // Deleting tail
          this.tail = current.prev;
          this.tail.next = null;
        } else {
          // Deleting middle node
          current.prev.next = current.next;
          current.next.prev = current.prev;
        }
        
        this._count--;
        return; // Delete only first occurrence
      }
      
      current = current.next;
    }
  }

  // Count elements in the list
  count() {
    return this._count;
  }
}

// Node class for doubly linked list
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
    this.prev = null;
  }
}

File: solutions/fortran/linked-list/1/linked_list.f90
module linked_list
  implicit none
  private

  type, public :: node_t
     integer                      :: value
     type(node_t), pointer :: next => null()
     type(node_t), pointer :: prev => null()
  end type node_t

  type, public :: list_t
     type(node_t), pointer :: head => null()
     type(node_t), pointer :: tail => null()
     integer               :: count = 0
  contains
     ! The following procedures are defined later
  end type list_t

  public :: new, push, pop, unshift, shift, length, delete, destroy

contains

  function new() result(list)
    type(list_t) :: list
    list%head => null()
    list%tail => null()
    list%count = 0
  end function new

  subroutine push(list, val)
    type(list_t), intent(inout) :: list
    integer, intent(in)         :: val
    type(node_t), pointer       :: new_node

    allocate(new_node)
    new_node%value = val
    new_node%next => null()
    new_node%prev => null()

    if (list%count == 0) then
       list%head => new_node
       list%tail => new_node
    else
       list%tail%next => new_node
       new_node%prev => list%tail
       list%tail => new_node
    end if
    list%count = list%count + 1
  end subroutine push

  function pop(list) result(val)
    type(list_t), intent(inout) :: list
    integer                     :: val
    type(node_t), pointer       :: old_tail

    if (list%count == 0) then
       val = -1   ! or handle error; the tests likely never pop empty list
       return
    end if

    old_tail => list%tail
    val = old_tail%value

    if (list%count == 1) then
       list%head => null()
       list%tail => null()
    else
       list%tail => old_tail%prev
       list%tail%next => null()
    end if

    deallocate(old_tail)
    list%count = list%count - 1
  end function pop

  subroutine unshift(list, val)
    type(list_t), intent(inout) :: list
    integer, intent(in)         :: val
    type(node_t), pointer       :: new_node

    allocate(new_node)
    new_node%value = val
    new_node%next => null()
    new_node%prev => null()

    if (list%count == 0) then
       list%head => new_node
       list%tail => new_node
    else
       new_node%next => list%head
       list%head%prev => new_node
       list%head => new_node
    end if
    list%count = list%count + 1
  end subroutine unshift

  function shift(list) result(val)
    type(list_t), intent(inout) :: list
    integer                     :: val
    type(node_t), pointer       :: old_head

    if (list%count == 0) then
       val = -1
       return
    end if

    old_head => list%head
    val = old_head%value

    if (list%count == 1) then
       list%head => null()
       list%tail => null()
    else
       list%head => old_head%next
       list%head%prev => null()
    end if

    deallocate(old_head)
    list%count = list%count - 1
  end function shift

  function length(list) result(n)
    type(list_t), intent(in) :: list
    integer                  :: n
    n = list%count
  end function length

  subroutine delete(list, val)
    type(list_t), intent(inout) :: list
    integer, intent(in)         :: val
    type(node_t), pointer       :: current

    current => list%head
    do while (associated(current))
       if (current%value == val) then
          ! Found node to delete
          if (associated(current%prev)) then
             current%prev%next => current%next
          else
             list%head => current%next   ! deleting head
          end if
          if (associated(current%next)) then
             current%next%prev => current%prev
          else
             list%tail => current%prev   ! deleting tail
          end if
          deallocate(current)
          list%count = list%count - 1
          exit
       end if
       current => current%next
    end do
    ! If list becomes empty, head/tail are nullified above
  end subroutine delete

  subroutine destroy(list)
    type(list_t), intent(inout) :: list
    type(node_t), pointer       :: current, next_node

    current => list%head
    do while (associated(current))
       next_node => current%next
       deallocate(current)
       current => next_node
    end do
    list%head => null()
    list%tail => null()
    list%count = 0
  end subroutine destroy

end module linked_list